gk/main.c

```diff
@@ -572,15 +671,29 @@ add_new_flow_from_policy(
 		return NULL;
 	}
 
-	/* Create a new flow entry. */
+	/*
+	 * We randomly drop entries to alleviate memory pressure
+	 * when the table is full.
+	 */
+hash_add_policy_flow:
 	ret = rte_hash_add_key_with_hash(
 		instance->ip_flow_hash_table,
  		&policy->flow, rss_hash_val);
-	if (ret < 0) {
+	if (ret == -EINVAL) {
 		RTE_LOG(ERR, HASH,
-			"The GK block failed to add new key to hash table in %s!\n",
+			"The GK block failed to add new key to hash table in %s with invalid parameters!\n",
 			__func__);
 		return NULL;
+	} else if (ret == -ENOSPC) {
+		RTE_LOG(WARNING, HASH,
+			"The GK block failed to add new key to hash table in %s due to lack of space!\n",
+			__func__);
+		ret = drop_flow_entry_randomly(
+			instance, gk_conf->flow_ht_size);
+		if (ret < 0)
+			return NULL;
+
+		goto hash_add_policy_flow;
```

Comment:
That's a good point! I have looked at the DPDK Hash table documentation (http://dpdk.org/doc/guides/prog_guide/hash_lib.html), and the hash table uses cuckoo hash method, in which there are two possible buckets (primary and secondary/alternative location) where that key can be stored in the hash. Probably we need to patch the hash table library to get the possible indexes for the given key, and iterate over the flow entry hash table according to the indexes, and find one to be replaced. Here is an issue: when the two buckets are full of entries that either granted or declined flows, do we still need to drop any entry?

Reply:
Come up with criteria to decide which entry to drop. Notice that it is fine to decide to drop the incoming entry instead of the entries already in the table. The code is dealing with overload, there is no perfect decision, we just need to make the best possible decision and ensure that we're not weakening Gatekeeper.

Reply:
Sounds good!